walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,46 +41,90 @@
                 }
             }
         }
+        List<TimeRange> availableTimes = getAvailableSlots(bookedTimes, request.getDuration());
+ 
+        //Get blocked and available times for optional people
+        for(Event event: events){
+            for(String attendee: request.getOptionalAttendees()){
+                if(event.getAttendees().contains(attendee)){
+                    bookedOptionalTimes.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+        List<TimeRange> availableOptionalTimes = getAvailableSlots(bookedOptionalTimes, request.getDuration());
 
-        if(bookedTimes.size() == 0){
-            slots.add(TimeRange.WHOLE_DAY);
+        // If no required attendee is available, then return no time slot available.
+        if(availableTimes.size() == 0){
             return slots;
         }
-
-        //Sort the booked times by start of day
-        Collections.sort(bookedTimes, TimeRange.ORDER_BY_START);
-
-        //Merge overlapping or nested timeslots into a single chunk of time
-        int i = 0;
-        while(i < bookedTimes.size() - 1){
-            if(bookedTimes.get(i).overlaps(bookedTimes.get(i+1)) || bookedTimes.get(i).contains(bookedTimes.get(i+1))){
-                int endTime = Math.max(bookedTimes.get(i).end(), bookedTimes.get(i+1).end());
-                bookedTimes.set(i, TimeRange.fromStartEnd(bookedTimes.get(i).start(), endTime, false));
-                bookedTimes.remove(i+1);
-            }else{
-                i++;
+ 
+        // Compare the timeslots available for required attendees and optional attendees and return the intersection.
+        for(int i = 0; i < availableTimes.size(); i++){
+            for(int j = 0; j < availableOptionalTimes.size(); j++){
```

Comment:
aren't both availableTimes and availableOptionalTimes sorted? In which case, could you could you just walk through the arrays using two counters and a while loop?

Reply:
Yes. I could do that. Would be a bit more complex because there might be overlaps in the available times. Also, one array may have time slots that start way later than the others. That way, we'll have to also keep track of the start and end times of the arrays at each index

We could have arrays that look like this: 
requiredTimes = [{8:00,8:30}, {9:00,11:00}, {13:00,:1500},{15:30,16:30}, {17:30,21:00}]
optionalTimes = [{14:00,18:00}, {20:00, 22:00}]

In this case, we would want our result to be [{15:30,16:30}]. Having one for-loop with two counters would need more if else checks to see which counter gets incremented where.

Depending on what you think, I can go ahead and implement it though

Reply:
Inside your for loop, you are checking
`if(availableTimes.get(i).end() < availableOptionalTimes.get(j).start())`
and
`if(availableTimes.get(i).contains(availableOptionalTimes.get(j)))`
My understanding is that you can have two counters (i,j) and update them smartly knowing that both arrays are sorted. If it's too complex then no need to do it.

Reply:
Implemented

Reply:
Resolved